### PR TITLE
Add a `get_if_exists` option for simpler creation of named actors

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Getting or creating a named actor is a common pattern, however it is somewhat esoteric in how to achieve this. Add a utility function and test that it doesn't cause any scary error messages.

`Actor.options(name="my_singleton", get_if_exists=True).remote(args)`

### CODE REVIEW

**Review by edoakes — COMMENTED** [review]
This is great; we have it littered throughout the serve code.

Should we have @pcmoritz sign off?

File: python/ray/actor.py
Comment (edoakes) [inline python/ray/actor.py:648]: is it worth logging a warning in this case?

Comment (ericl) [reply python/ray/actor.py:648]: It's expected, so we shouldn't log anything (people have complained about this before when we force-logged warnings in core worker).

Comment (edoakes) [reply python/ray/actor.py:648]: fair point; ideally we'd use [warnings](https://docs.python.org/3/library/warnings.html) everywhere so people could suppress things as desired

**Review by rkooo567 — COMMENTED** [review]
Code-wise, looks good to me. Will wait for API approval from @pcmoritz (so I will defer approval to him)

File: python/ray/tests/test_get_or_create_actor.py
Comment (scv119) [inline python/ray/tests/test_get_or_create_actor.py:9]: should add some concurrent test to ensure no surprise?

**Review by scv119 — APPROVED** [review]

**Comment by scv119** [comment]
ah this need api-change approval.. probably shouldn't approve by now.

File: python/ray/tests/test_get_or_create_actor.py
Comment (ericl) [reply python/ray/tests/test_get_or_create_actor.py:9]: Yep, the verbose output one is actually a concurrency test. In the past, this pattern sometimes generated error logs under specific concurrent race conditions.

**Comment by pcmoritz** [comment]
The functionality here makes a lot of sense and we should totally implement it! The API seems a little ugly of course, partly because of the asymmetry with `ray.get_actor`, and partly because it tacks on a new method besides .remote to the actor. I think because this is part of the core API it is worth thinking a bit more about the alternatives. The following options come to mind, all of them are suboptimal:

- Introduce a flag for actor creation now like Actor.options(name="name", get_or_create=True) -- this is more consistent with the current way but also the syntax possibly makes it less clear what is going on. Together with the proposal in the PR these are the main contenders in my mind at the moment.
- Have ray.get_or_create_actor symmetrically to ray.get_actor -- this is more symmetric but it pollutes the public namespace further and also is inconsistent with how we create named actors today, so I'd say it should not be the recommended way.

Going forward it seems to me that ideally get_or_create should be the default/strongly suggested way for people to do this. How can we make that the case without breaking the API? Any ideas?

It is possible that I'm trying to square the circle here, but I wonder if other people have more ideas.

**Comment by ericl** [comment]
Merged, thanks! Also filed the followup. I think it's doable but may need C++ changes.

**Comment by doliynyk-nference** [comment]
@ericl What version of ray is this on? I would like to use this feature but when I try it on ray 1.12.1 I get `TypeError: options() got an unexpected keyword argument 'get_if_exists'`

**Pull request merged**
